dowhy/gcm/independence_test/ablation.py

```diff
@@ -0,0 +1,107 @@
+""" Ablation based (conditional) independence test. Testing independence via regression, i.e. if a variable has
+information about another variable, then they are dependent.
+"""
+from typing import Callable, Optional, Union, List
+
+import numpy as np
+from sklearn.kernel_approximation import Nystroem
+from sklearn.preprocessing import scale
+
+from dowhy.gcm.stats import quantile_based_fwer, estimate_f_test_p_value
+from dowhy.gcm.util.general import shape_into_2d, apply_one_hot_encoding, fit_one_hot_encoders
+
+
+def ablation_based(X: np.ndarray,
+                   Y: np.ndarray,
+                   Z: Optional[np.ndarray] = None,
+                   num_components_all_inputs: int = 40,
+                   num_runs: int = 20,
+                   p_value_adjust_func: Callable[[Union[np.ndarray, List[float]]], float] = quantile_based_fwer,
+                   f_test_samples_ratio: Optional[float] = 0.3,
+                   max_samples_per_run: int = 10000) -> float:
+    """ The main idea is that if X and Y are dependent, then X should help in predicting Y. If there is no dependency,
+    then X should not help. When Z is given, the idea remains the same, but here X and Y are conditionally independent
+    given Z if X does not help in predicting Y when knowing Z. This is, X has not additional information about Y given
+    Z. In the pairwise case (Z is not given), the performances between predicting Y based on X and predicting Y by
+    returning its mean (the best estimator without any inputs) are compared.
+
+    Here, we use the Nystroem approach to approximate a kernel map of the inputs that serves as new input features.
+    These new features allow to model complex non-linear relationships. In case of categorical data, we first apply an
+    one-hot-encoding and then map it into the kernel feature space. Afterwards, we use linear regression as a prediction
+    model based on the non-linear input features. The idea is then the same as in Granger causality, where we
+    apply a f-test to see if the additional input features significantly help in predicting the target or not.
+
+    Note: As compared to kernel_based, this method is quite fast and provides reasonably well results. However, there
+    are types of dependencies that this test can not detect. For instance, if X determines the variance of Y, then this
+    can not be captured. For these more complex dependencies, consider using the kernel_based independence test instead.
+
+    This test is motivated by Granger causality, the approx_kernel_based test and the following paper:
+        K Chalupka, P Perona, F. Eberhardt. *Fast Conditional Independence Test for Vector Variables with Large Sample Sizes*. arXiv:1804.02747, 2018.
+
+    :param X: Input data for X.
+    :param Y: Input data for Y.
+    :param Z: Input data for Z.
+    :param num_components_all_inputs: Number of kernel features when combining X and Z. Half of the number is used when
```

Comment:
compared with Z alone---when is this done? I thought if Z is optional, we'd be only using X features